### PR TITLE
UCT/DC: preparations for the hw tag matching

### PR DESCRIPTION
**Pull request by alex-mikheev**
@yosefe please take a look

### CODE REVIEW

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4507/ for details (Mellanox internal link).

File: src/uct/ib/rc/verbs/rc_verbs_common.c [hunk 1]
```diff
@@ -23,9 +23,82 @@ ucs_config_field_t uct_rc_verbs_iface_common_config_table[] = {
...
+  {"TM_SYNC_RATIO", "0.5",
+   "Maximal portion of the tag matching list which can be canceled without requesting\n"
+   "a completion.",
+   ucs_offsetof(uct_rc_verbs_iface_common_config_t, tm.sync_ratio), UCS_CONFIG_TYPE_DOUBLE},
+#endif
+
   {NULL}
 };
 
+static void uct_rc_verbs_iface_tag_query(uct_rc_verbs_iface_common_t *iface,
```
Comment (yosefe) [inline src/uct/ib/rc/verbs/rc_verbs_common.c]: uct_rc_verbs_iface_common_tag_query

File: src/uct/ib/rc/verbs/rc_verbs_common.c [hunk 2]
```diff
@@ -23,9 +23,82 @@ ucs_config_field_t uct_rc_verbs_iface_common_config_table[] = {
...
+                                         uct_ib_iface_t *ib_iface,
+                                         uct_iface_attr_t *iface_attr)
+{
+#if IBV_EXP_HW_TM
+    uct_ib_device_t *dev     = uct_ib_iface_device(ib_iface);
+    unsigned eager_hdr_size  = sizeof(struct ibv_exp_tmh);
+
+    if (iface->tm.enabled) {
```
Comment (yosefe) [inline src/uct/ib/rc/verbs/rc_verbs_common.c]: if (!enabled) return

File: src/uct/ib/rc/verbs/rc_verbs_common.c [hunk 3]
```diff
@@ -96,24 +172,179 @@ ucs_status_t uct_rc_verbs_iface_prepost_recvs_common(uct_rc_iface_t *iface,
...
+    uint32_t cap_flags   = IBV_DEVICE_TM_CAPS(dev, capability_flags);
+    struct ibv_exp_tmh tmh;
+    int tm_supported;
+
+    /* DC is not supported yet */
+    tm_supported = is_dc ? 0 : (cap_flags & IBV_EXP_TM_CAP_RC);
+
+    iface->tm.enabled = (config->tm.enable && tm_supported);
+
+    if (iface->tm.enabled) {
```
Comment (yosefe) [inline src/uct/ib/rc/verbs/rc_verbs_common.c]: if (!enabled) { *rx_hdr_len=... ...   ;   return }

File: src/uct/ib/rc/verbs/rc_verbs_common.c [hunk 4]
```diff
@@ -96,24 +172,179 @@ ucs_status_t uct_rc_verbs_iface_prepost_recvs_common(uct_rc_iface_t *iface,
...
+                            uct_rc_iface_config_t *rc_config)
+{
+#if IBV_EXP_HW_TM
+    struct ibv_exp_create_srq_attr srq_init_attr;
+    ucs_status_t status;
+    int sync_ops_count;
+    int rx_hdr_len;
+    uct_ib_md_t *md = ucs_derived_of(rc_iface->super.super.md, uct_ib_md_t);
+
+    if (iface->tm.enabled) {
```
Comment (yosefe) [inline src/uct/ib/rc/verbs/rc_verbs_common.c]: if (!enabled) return OK

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4512/ for details (Mellanox internal link).

**Comment by yosefe** [comment]
bot:mlx:retest

**Pull request merged**